src/files/BinaryFile.cpp

```diff
@@ -0,0 +1,751 @@
+        // make sure the file fits inside mmap_size_
+        while (file_size_ > mmap_size_) {
+            mmap_size_ *= 2;
+        }
+
+        offset_ = file_size_;
+        total_written_size_ = file_size_;
+    } else {
+        unreachable();
+    }
+
+    mmap_data_ = static_cast<char*>(mmap(
+        nullptr, mmap_size_, mmap_prot_, MAP_SHARED, file_descriptor_, 0
+    ));
+
+    if (mmap_data_ == MAP_FAILED) {
+        throw file_error("mmap failed for '{}': {}", this->path(), std::strerror(errno));
+    }
+#else
+    const char* open_mode;
+    if (mode == Mode::READ) {
+        open_mode = "rb";
+    } else if (mode == Mode::APPEND) {
+        open_mode = "r+b";
+    } else if (mode == Mode::WRITE) {
+        open_mode = "wb";
+    } else {
+        unreachable();
+    }
+
+    file_ = std::fopen(this->path().c_str(), open_mode);
+    if (file_ == nullptr) {
+        throw file_error(
+            "could not open file at '{}': {}", this->path(), std::strerror(errno)
+        );
+    }
+
+    if (mode == Mode::APPEND) {
+        fseek64(file_, 0, SEEK_END);
+    } else {
+        this->seek(0);
+    }
+
+#endif
+}
+
+BinaryFile::~BinaryFile() {
+#if CHEMFILES_BINARY_FILE_USE_MMAP
+    if (mmap_data_ != nullptr) {
+        msync(mmap_data_, mmap_size_, MS_SYNC);
+        munmap(mmap_data_, mmap_size_);
+    }
+
+    if (file_descriptor_ != -1) {
+        if (this->mode() != Mode::READ) {
+            auto status = ftruncate(file_descriptor_, static_cast<off64_t>(total_written_size_));
+            if (status != 0) {
+                warning(
+                    "binary file writer",
+                    "failed to resize file to it's actual size when closing: {}",
+                    std::strerror(errno)
+                );
+            }
+        }
+
+        close(file_descriptor_);
+    }
+
+    file_descriptor_ = -1;
+    total_written_size_ = 0;
+    mmap_data_ = nullptr;
+    file_size_ = 0;
+    mmap_size_ = 0;
+    mmap_prot_ = 0;
+    page_size_ = 0;
+    offset_ = 0;
+#else
+    if (this->mode() != Mode::READ) {
+        std::fflush(file_);
+    }
+
+    std::fclose(file_);
+    file_ = nullptr;
+#endif
+}
+
+BinaryFile& BinaryFile::operator=(BinaryFile&& other) {
+    this->~BinaryFile();
+
+    File::operator=(std::move(other));
+
+#if CHEMFILES_BINARY_FILE_USE_MMAP
+    std::swap(this->file_descriptor_, other.file_descriptor_);
+    std::swap(this->total_written_size_, other.total_written_size_);
+    std::swap(this->mmap_data_, other.mmap_data_);
+    std::swap(this->file_size_, other.file_size_);
+    std::swap(this->mmap_size_, other.mmap_size_);
+    std::swap(this->mmap_prot_, other.mmap_prot_);
+    std::swap(this->page_size_, other.page_size_);
+    std::swap(this->offset_, other.offset_);
+#else
+    std::swap(this->file_, other.file_);
+#endif
+
+    return *this;
+}
+
+void BinaryFile::read_char(char* data, size_t count) {
+#if CHEMFILES_BINARY_FILE_USE_MMAP
+    std::memcpy(data, mmap_data_ + offset_, count);
```

Comment:
One can read beyond the end of a mmap which returns junk values. Might be better to throw an exception when reaching EOF. Should there be test case for that?
```suggestion
    if (offset_ + count > mmap_size_) {
        throw file_error(
            "failed to read {} bytes from the file at '{}': mmap out of bounds",
            count, this->path()
        );
    }
    std::memcpy(data, mmap_data_ + offset_, count);
```

Reply:
reading past the end would probably segfault, but giving a good error message is a good idea. I'll triple check that it does not degrade performances.